projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/intrinsics/c/LLVMSignal.java

```diff
@@ -0,0 +1,425 @@
+     * timing issues when calling multiple signals in a defined sequence, or when a program has to
+     * wait until the signal was handled (which is not guaranteed because of the asynchronous
+     * behavior in our implementation).
+     */
+    private static final class LLVMSignalHandler implements SignalHandler, LLVMThreadNode {
+
+        private final Signal signal;
+        private final LLVMFunctionDescriptor function;
+        private final LLVMContext context;
+        private RootCallTarget callTarget;
+        private final LLVMStack stack = new LLVMStack();
+
+        private final Lock lock = new ReentrantLock();
+        private final AtomicBoolean isRunning = new AtomicBoolean(false);
+
+        @TruffleBoundary
+        private LLVMSignalHandler(Signal signal, LLVMFunctionDescriptor function) throws IllegalArgumentException {
+            this.signal = signal;
+            this.function = function;
+            LLVMFunctionNode functionNode = LLVMFunctionLiteralNodeGen.create(function);
+
+            LLVMAddressLiteralNode signalStack = new LLVMAddressLiteralNode(stack.allocate(TMP_SIGNAL_STACK_SIZE_BYTE));
+            LLVMI32LiteralNode sigNumArg = new LLVMI32LiteralNode(signal.getNumber());
+            LLVMExpressionNode[] args = {signalStack, sigNumArg};
+
+            context = LLVMLanguage.INSTANCE.findContext0(LLVMLanguage.INSTANCE.createFindContextNode0());
+
+            LLVMUnresolvedCallNode callNode = new LLVMUnresolvedCallNode(functionNode, args, LLVMRuntimeType.VOID, context);
+
+            callTarget = Truffle.getRuntime().createCallTarget(
+                            new LLVMFunctionStartNode(callNode,
+                                            new LLVMNode[]{},
+                                            new LLVMNode[]{},
+                                            SourceSection.createUnavailable("", null),
+                                            new FrameDescriptor(), ""));
+
+            lock.lock();
+            try {
+                if (function.equals(LLVM_SIG_DFL)) {
+                    Signal.handle(signal, SignalHandler.SIG_DFL);
+                    stack.free();
+                } else if (function.equals(LLVM_SIG_IGN)) {
+                    Signal.handle(signal, SignalHandler.SIG_IGN);
+                    stack.free();
+                } else {
+                    Signal.handle(signal, this);
+                }
+
+                // only when we reach this point, the signal handler was registered successfully
+                isRunning.set(true);
+                context.registerThread(this);
+            } catch (IllegalArgumentException e) {
+                stack.free();
+                throw e;
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        public boolean isRunning() {
+            return isRunning.get();
+        }
+
+        @Override
+        @TruffleBoundary
+        protected void finalize() throws Throwable {
```

Comment:
Finalizers are not guaranteed to be executed. Does your approach still work if they aren't?